### PR TITLE
Reduce spurious replication catchup

### PR DESCRIPTION
**Pull request by erikjohnston**
The `POSITION` command includes two tokens, the "prev" and "new" tokens. The "new" token is what we should set the current token for the instance to be, and the "prev" token was where it was previously (i.e. the instance produced no rows between the "prev" and "new" tokens).

This means that we can safely relax the condition to only check for missing updates if the current token for the writer is strictly less than the "prev" token.

A common case is receiving something like `POSITIONS event_persister1 events 15 16`, which indicates the writer advanced their events stream token from 15 to 16 (usually as a response to another instance persisting an event with stream ID 16). Usually, the other instances would already have inferred that `event_persister1` must have advanced their token, so the `current_token` would already be set to 16, thus triggering a lookup for missing updates between tokens 16 (`cmd.new_token`) and 16 (`current_token`), which is silly.

We also move the logging back into the loop so that we don't log in the common case.

### CODE REVIEW

File: synapse/replication/tcp/handler.py
Comment (DMRobertson) [inline synapse/replication/tcp/handler.py:616]:
```suggestion
        # If the incoming position is at least our current position then we're 
        # up to date and there's nothing to do. Otherwise, fetch all updates
        # between then and now.
```

File: synapse/replication/tcp/handler.py
Comment (DMRobertson) [inline synapse/replication/tcp/handler.py:620]:
```suggestion
        # Note: We also have to check that `current_token` is at least the
        # previous position, to handle the case where the stream gets "reset" 
        # (e.g. for `caches` and `typing` after the writer's restart).
```

File: synapse/replication/tcp/handler.py
Comment (DMRobertson) [reply synapse/replication/tcp/handler.py:616]: Example: I think the stream is at position 10.
I receive `POSITION 4 8`.
Uh oh, I might have missed stream entries 5, 6, 7 and 8. Need to rescan for them.

File: synapse/replication/tcp/handler.py
Comment (DMRobertson) [reply synapse/replication/tcp/handler.py:620]: Example: I think the stream is at position 10.
I receive `POSITION 12 15`.
Uh oh, I might have missed stream entries 11, 12, 13, 14, and 15. Need to rescan for them.

File: synapse/replication/tcp/handler.py
Comment (DMRobertson) [reply synapse/replication/tcp/handler.py:616]: Example: I think the stream is at position 10.
I receive `POSITION 12 15`.
Uh oh, I might have missed stream entries 11, 12, 13, 14, and 15. Need to rescan for them.

File: synapse/replication/tcp/handler.py
Comment (DMRobertson) [reply synapse/replication/tcp/handler.py:620]: Example: I think the stream is at position 10.
I receive `POSITION 12 15`.
Uh oh, I might have missed stream entries 11, 12, 13, 14, and 15. Need to rescan for them.

**Review by DMRobertson — APPROVED** [review]
Is my understanding correct? LGTM if so.

File: synapse/replication/tcp/handler.py
Comment (DMRobertson) [inline synapse/replication/tcp/handler.py:621]: Sanity check: are the inequalities correct here? E.g. should one of them be `<` instead of `<=`?

File: synapse/replication/tcp/handler.py
Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:616]: Ah, no. `POSITION 4 8` means that the new token for that instance should be set to 8, and nothing has happened between 4 and 8 from that instance. (The reason for this is that normally we'd just get a stream of  `RDATA`  from the instance, which implicitly bumps the token. If for some reason the token has advanced *without* any updates to send, it sends a `POSITION`, hence why it gives the range where things have *not* happened).

Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:616]:
```suggestion
        # If the incoming previous position is less than our current position
	    # then we're  up to date and there's nothing to do. Otherwise, fetch
	    # all updates between then and now.
```

Maybe?

File: synapse/replication/tcp/handler.py
Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:620]: > Uh oh, I might have missed stream entries 11, 12, 13, 14, and 15. Need to rescan for them.

You might have missed entries 11 and 12.

Comment (clokep) [reply synapse/replication/tcp/handler.py:620]: Does `POSITION` need some better documentation somewhere (either https://matrix-org.github.io/synapse/develop/development/synapse_architecture/streams.html or maybe in the code?) describing how it works? I think I was also under the impression that it told you were it had updates, not where it jumped to *without* updates.

Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:620]: https://matrix-org.github.io/synapse/develop/tcp_replication.html#position-s probably needs to be updated too

Comment (clokep) [reply synapse/replication/tcp/handler.py:620]: (Not blocking this PR, but would appreciate a follow-up or at least an issue.)

Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:620]: https://github.com/matrix-org/synapse/issues/16560 since its a Friday afternoon and I'm liable to forget to follow up otherwise

Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:620]:
```suggestion
        # Note: We also have to check that `current_token` is at least the
        # new position, to handle the case where the stream gets "reset" 
        # (e.g. for `caches` and `typing` after the writer's restart).
```

File: synapse/replication/tcp/handler.py
Comment (clokep) [inline synapse/replication/tcp/handler.py:620]: I think?

```suggestion
        # Note: We also have to check that `current_token` is at most the
        # new position, to handle the case where the stream gets "reset" 
        # (e.g. for `caches` and `typing` after the writer's restart).
```

Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:620]: Yes :facepalm:

**Review by clokep — APPROVED** [review]

**Pull request merged**
